control/metrics.go

```diff
@@ -202,120 +230,197 @@ func newMetricCatalog() *metricCatalog {
-// an asterisk or tuple (refer to query support)
-func (mc *metricCatalog) GetQueriedNamespaces(ns core.Namespace) ([]core.Namespace, error) {
-	mc.mutex.Lock()
-	defer mc.mutex.Unlock()
-
-	// get metric key (might contain wildcard(s))
-	wkey := ns.Key()
+// getMatchedMetricTypes returns all matched metric types stored under the key 'mkey' in map 'mTree'
+func (mc *metricCatalog) getMatchedMetricTypes(mkey mapKey) ([]*metricType, error) {
+	mts := mc.mTree[mkey]
+	if len(mts) == 0 {
+		return nil, errorMetricNotFound(mkey.metricNamespace(), mkey.metricVersion())
+	}
 
-	return mc.matchedNamespaces(wkey)
+	return mts, nil
 }
 
-// MatchQuery matches given 'ns' which could contain an asterisk or a tuple and add them to matching map under key 'ns'
-// The matched metrics namespaces are also returned (as a []core.Namespace)
-func (mc *metricCatalog) MatchQuery(ns core.Namespace) ([]core.Namespace, error) {
-	mc.mutex.Lock()
-	defer mc.mutex.Unlock()
+// findTupleSubmatch returns all matched combination of queried tuples
+// where as a tuple there is a mean of `(a|b)`
+func findTupleSubmatch(ns []string) [][]string {
+	/*
+		Example:
+		findTupleSubmatch([]string{"intel", "mock", "(host0|host1)", "(baz|bar)"})
+
+		would return four slices:
+			[]string{"intel", "mock", "host0", "baz"}
+			[]string{"intel", "mock", "host1", "baz"}
+			[]string{"intel", "mock", "host0", "bar"}
+			[]string{"intel", "mock", "host1", "bar"}
+	*/
+
+	numOfPossibleCombinations := 1
+	matchedItems := make(map[int][]string)
+
+	for index, n := range ns {
+		match := []string{}
+
+		if strings.ContainsAny(n, "*") {
+			// an asterisk covers all tuples cases
+			match = []string{"*"} // to avoid retrieving the same metric more than once
+		} else {
+			// a tuple is equivalent to regexp token (e.q match either a or b: '(a|b)')
+			// so it provides regular expression by itself
+			regex := regexp.MustCompile(n)
+			match = regex.FindAllString(n, -1)
+		}
 
-	// get metric key (might contain wildcard(s))
-	wkey := ns.Key()
+		if match == nil {
+			continue
+		}
 
-	// adding matched namespaces to map
-	mc.addItemToMatchingMap(wkey)
+		matchedItems[index] = append(matchedItems[index], match...)
 
-	return mc.matchedNamespaces(wkey)
-}
+		// number of possible combinations increases N=len(match) times
+		numOfPossibleCombinations = numOfPossibleCombinations * len(match)
 
-func convertKeysToNamespaces(keys []string) []core.Namespace {
-	// nss is a slice of slices which holds metrics namespaces
-	nss := []core.Namespace{}
-	for _, key := range keys {
-		ns := getMetricNamespace(key)
-		if len(ns) != 0 {
-			nss = append(nss, ns)
+	}
+
+	//prepare two dimensional slice representing namespaces
+	nss := make([][]string, numOfPossibleCombinations)
+
+	for index := 0; index < len(ns); index++ {
+		items := matchedItems[index]
+		fillThreshold := len(items)
+
+		for i := 0; i < numOfPossibleCombinations; i++ {
```

Comment:
The minimum number of possible combination is 1 what means that even the incoming **ns** does not contain a tuple, the result ([][]string) has to contain that one **ns**.